### PR TITLE
Fix unintended checkbox (state icon) state change  in multi selection…

### LINKED ISSUE #2902: Unintended checkbox (state icon) state change  in multi selection case with SHIFT+LeftMouseClick when clicked not on the first column
* .NET Core Version:

The issue was found in .NET Framework, but I believe .NET Core inherited it as the identified source code cause is the same.

* Have you experienced this same bug with .NET Framework?:

 Yes

**Problem description:**

System.Windows.Forms.ListView.WmNmClick implementation contains:
https://github.com/dotnet/winforms/blob/8af8e84c51a0740c339a292263c94d25b46f5324/src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs#L5547-L5551

However the `LVHT.ONITEMSTATEICON` flag is not valid if clicked not on the first column. The method should call `LVM.SUBITEMHITTEST` instead and check subitem like:
```cs
int displayIndex = (int)User32.SendMessageW(this, (User32.WM)LVM.SUBITEMHITTEST, IntPtr.Zero, ref lvhi);
if (displayIndex == -1 || lvhi.iSubItem != 0 || (lvhi.flags & LVHT.ONITEMSTATEICON) == 0)
{
    return;
}
```
The bug leads to checkboxes (state icons) change in multi selection case with SHIFT+LeftMouseClick when clicked not on the first column:

Demo

**Expected behavior:**

Checkboxes are not affected by the click.

**Minimal repro:**

The [zip](https://developercommunity.visualstudio.com/storage/attachments/29088-winformslistviewbug.zip) contains repro VS project for .NET Framework.

### PR DESCRIPTION
**Pull request by JarLob**
… case with SHIFT+LeftMouseClick when clicked not on the first column

Fixes #2902

## Proposed changes

Use correct windows message and ignore the click if not on the first column.

## Regression? 

- No

## Risk

Low, if any customer hooked the listview to workaround the bug they should not be affected by the fix.

## Test methodology 

Not sure if a test can be written. I looked at existing test, but didn't find any that would send clicks to a list view. Suggestions are welcome.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2903)

### CODE REVIEW

**Comment by hughbe** [comment]
Could possibly write an integration test?

**Comment by JarLob** [comment]
Any pointers to what I could use as an example?

**Comment by RussKie** [comment]
@Vino-Wang is this something we can test via MAUI?

**Comment by RussKie** [comment]
@JarLob we have integration tests (that we call MAUI) located at https://github.com/dotnet/winforms/tree/master/src/System.Windows.Forms/tests/IntegrationTests/MauiTests.
So it should be possible to create a new test project for testing the ListView, and specifically the problem you have kindly offered to fix.

A mouse cursor can be moved programmatically via `Cursor.Position` API (that calls underlying Win32 API):
https://github.com/dotnet/winforms/blob/b666dc7a94d8ac87a7d300cfb4fa86332fb79bae/src/System.Windows.Forms/src/System/Windows/Forms/Cursor.cs#L174-L182
You may also need to translate relative coordinates to absolute.


To simulate keyboard events and mouse clicks you could use [`SendInput`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-input) function
https://github.com/dotnet/winforms/blob/b666dc7a94d8ac87a7d300cfb4fa86332fb79bae/src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.SendInput.cs#L12-L20

Shout if you get stuck, we'll try to help.

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs [hunk 1]
```diff
@@ -579,75 +579,6 @@ patternId switch
                     _ => base.IsPatternSupported(patternId)
                 };
 
-            public void RaiseMouseClick(int x, int y)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs]: ❗️ We must not make changes to the production code in order to accommodate testing, especially in completely unrelated areas (i.e. we're fixing a bug in `ListView` yet we are modifying `MonthCalendar.MonthCalendarAccessibleObject` and exposing various libraries to the test project).

It is completely acceptable to make a copy of the code and place it in the test project, if necessary.

**Comment by RussKie** [comment]
@Vino-Wang @AdamYoblick please review the tests, thank you

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs (hunk 1 above)
Comment (JarLob) [reply src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs]: As you say. Reverted the change.
P.S. As a side note mouse click sending functionality is not an integral part of MonthCalendar and was previously extracted (without any modification) to a separate _internal_ class.

**Comment by RussKie** [comment]
Thank you

**Pull request merged**